python/packages/core/agent_framework/azure/_search_provider.py

```diff
@@ -0,0 +1,706 @@
+except ImportError:
+    _agentic_retrieval_available = False
+
+if sys.version_info >= (3, 11):
+    from typing import Self  # pragma: no cover
+else:
+    from typing_extensions import Self  # pragma: no cover
+
+if sys.version_info >= (3, 12):
+    from typing import override  # type: ignore # pragma: no cover
+else:
+    from typing_extensions import override  # type: ignore[import] # pragma: no cover
+
+
+class AzureAISearchContextProvider(ContextProvider):
+    """Azure AI Search Context Provider with hybrid search and semantic ranking.
+
+    This provider retrieves relevant documents from Azure AI Search to provide context
+    to the AI agent. It supports two modes:
+
+    - **semantic** (default): Fast hybrid search combining vector and keyword search
+      with semantic reranking. Suitable for most RAG use cases.
+    - **agentic**: Slower multi-hop reasoning across documents using Knowledge Bases.
+      Use only for complex queries requiring cross-document reasoning.
+
+    Examples:
+        Semantic hybrid search (recommended for most cases):
+
+        .. code-block:: python
+
+            from agent_framework import ChatAgent
+            from agent_framework_azure_ai import AzureAIAgentClient, AzureAISearchContextProvider
+            from azure.identity.aio import DefaultAzureCredential
+
+            # Create context provider with semantic hybrid search
+            search_provider = AzureAISearchContextProvider(
+                endpoint="https://mysearch.search.windows.net",
+                index_name="my-index",
+                credential=DefaultAzureCredential(),
+                mode="semantic",  # Fast hybrid + semantic ranker (default)
+            )
+
+            # Use with agent
+            async with (
+                AzureAIAgentClient() as client,
+                ChatAgent(
+                    chat_client=client,
+                    context_providers=[search_provider],
+                ) as agent,
+            ):
+                response = await agent.run("What is in the documents?")
+
+        Agentic retrieval for complex queries:
+
+        .. code-block:: python
+
+            # Use agentic mode for multi-hop reasoning (slower)
+            search_provider = AzureAISearchContextProvider(
+                endpoint="https://mysearch.search.windows.net",
+                index_name="my-index",
+                credential=DefaultAzureCredential(),
+                mode="agentic",  # Multi-hop reasoning
+                azure_ai_project_endpoint="https://myproject.services.ai.azure.com",
+                model_deployment_name="gpt-4o",
+                knowledge_base_name="my-knowledge-base",  # Required for agentic mode
+            )
+    """
+
+    _DEFAULT_SEARCH_CONTEXT_PROMPT = "Use the following context to answer the question:"
+
+    def __init__(
+        self,
+        endpoint: str,
+        index_name: str,
+        credential: AzureKeyCredential | AsyncTokenCredential,
```

Comment:
let's also allow folks to pass in a key directly, and then we create the AzureKeyCredential in code.